### PR TITLE
V2 stream xcaching

### PR DESCRIPTION
**Pull request by 0xakk0r0kamui**

### CODE REVIEW

File: chain/getblocks.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	g.newHeight = g.req.Heights[0] - 1
+	g.blkRecv = make(chan common.ExpectedBlk, common.MaxBlocksPerRequest)
+	g.updateNewHeight()
+	return g
+}
+
+func (g *BlkGetter) Cancel() {
+	for range g.blkRecv {
+	}
+}
```
Comment (0xbunyip) [inline chain/getblocks.go]: remove unused method

File: chain/getblocks.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import (
+	context "context"
+	"highway/common"
+	"highway/proto"
+)
+
+type BlkGetter struct {
+	waiting   map[uint64][]byte
+	newBlk    chan common.ExpectedBlk
+	newHeight uint64
```
Comment (0xbunyip) [inline chain/getblocks.go]: newHeight is only used in listenCommingBlock, we can define it as a local var in that method

File: chain/getblocks.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+func (g *BlkGetter) checkWaitingBlk() bool {
+	if data, ok := g.waiting[g.newHeight]; ok {
+		g.blkRecv <- common.ExpectedBlk{
+			Height: g.newHeight,
+			Data:   data,
+		}
+		delete(g.waiting, g.newHeight)
+		g.updateNewHeight()
+		return true
+	}
```
Comment (0xbunyip) [inline chain/getblocks.go]: we should return a range of block when checking for waiting block
e.g.: have block [2, 3, 4] => receive 1 => return block [1, 2, 3, 4]

File: chain/client.go [hunk 4]
```diff
@@ -490,6 +347,35 @@ func (cc *ClientConnector) GetServiceClient(peerID peer.ID) (proto.HighwayServic
 	return client, nil
 }
 
+func (cc *ClientConnector) GetGRPCConn(peerID peer.ID) (*grpc.ClientConn, error) {
```
Comment (0xbunyip) [inline chain/client.go]: remove unused method, pls

File: chain/memcache.go [hunk 5]
```diff
@@ -39,22 +39,22 @@ func getKeyByHeight(req GetBlockByHeightRequest, h uint64) string {
...
-		key := getKeyByHeight(req, h)
-		if b, ok := cache.cacher.Get(key); ok {
-			if block, ok := b.([]byte); ok {
-				blocks[i] = block
-			}
-		}
-	}
-	return blocks, nil
-}
+// func (cache *MemCache) GetBlockByHeight(_ context.Context, req GetBlockByHeightRequest, heights []uint64) ([][]byte, error) {
```
Comment (0xbunyip) [inline chain/memcache.go]: remove commented code, pls

File: chain/getblocks.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+					missing = append(missing, req.Heights[i:]...)
+					break
+				}
+			}
+		} else {
+			for height := last + 1; height <= req.Heights[len(req.Heights)-1]; height++ {
+				missing = append(missing, height)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (0xakk0r0kamui) [reply chain/getblocks.go]: oke

File: chain/clientstreamutils.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+func (c *Client) StreamBlkByHeight(
+	ctx context.Context,
+	req RequestBlockByHeight,
+	blkChan chan common.ExpectedBlk,
+) error {
+	logger := Logger(ctx)
+	defer close(blkChan)
+	logger.Infof("[stream] Server call Client: Start stream request %v", req)
+	sc, _, err := c.getClientWithBlock(ctx, int(req.GetFrom()), req.GetHeights()[len(req.GetHeights())-1])
+	if sc == nil {
```
[the remark this answers is not in the document]
Comment (0xakk0r0kamui) [reply chain/clientstreamutils.go]: done

File: chain/clientstreamutils.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	idx := 0
+	blkData := new(proto.BlockData)
+	for blkHeight < heights[len(heights)-1] {
+		if req.GetSpecific() {
+			blkHeight = heights[idx]
+			idx++
+		} else {
+			blkHeight++
+		}
+		if err == nil {
```
[the remark this answers is not in the document]
Comment (0xakk0r0kamui) [reply chain/clientstreamutils.go]: Please re-visit this code, I modified for easy understanding.

File: chain/getblocks.go (hunk 1 above)
Comment (0xakk0r0kamui) [reply chain/getblocks.go]: Forgot calling this func

File: chain/getblocks.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+		if blk.Height < g.newHeight {
+			continue
+		}
+		if blk.Height == g.newHeight {
+			g.blkRecv <- blk
+			g.updateNewHeight()
+		} else {
+			g.waiting[blk.Height] = blk.Data
+		}
+		for hasNewBlk := g.checkWaitingBlk(); hasNewBlk == true; {
```
Comment (0xbunyip) [inline chain/getblocks.go]: we can put for loop inside checkWaitingBlk()
then we wouldn't need to loop at line 71

File: chain/clientstreamutils.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
+package chain
+
+import (
+	context "context"
+	"highway/common"
+	"highway/proto"
+)
+
+func (c *Client) StreamBlkByHeight(
```
Comment (0xbunyip) [inline chain/clientstreamutils.go]: we should call capBlocksPerRequest() before serving the request

**Pull request merged**